languages/rust/oso/src/host/to_polar.rs

```diff
@@ -18,25 +21,82 @@ use super::Host;
 ///
 /// For non-primitive types, the instance will be stored
 /// on the provided `Host`.
-pub trait ToPolar {
-    fn to_polar_value(&self, host: &mut Host) -> Value;
+pub trait ToPolar: Send + Sync + Sized + 'static {
+    fn to_polar_value(self, host: &mut Host) -> Value {
+        let instance = Instance::new(self);
+        let instance = host.cache_instance(instance, None);
+        Value::ExternalInstance(ExternalInstance {
+            constructor: None,
+            repr: Some(std::any::type_name::<Self>().to_owned()),
+            instance_id: instance,
+        })
+    }
 
-    fn to_polar(&self, host: &mut Host) -> Term {
+    fn to_polar(self, host: &mut Host) -> Term {
         Term::new_from_ffi(self.to_polar_value(host))
     }
 }
 
+impl<C: crate::PolarClass + Send + Sync> ToPolar for C {
+    fn to_polar_value(self, host: &mut Host) -> Value {
+        let instance = Instance::new(self);
+        let instance = host.cache_instance(instance, None);
+        if host.get_class_from_type::<Self>().is_none() {
+            let class = Self::get_polar_class();
+            let name = Symbol(class.name.clone());
+            let _ = host.cache_class(class, name);
```

Comment:
Let's add a log here.  We are also silently swallowing this error. Let's at least unwrap & add a todo.  `to_polar_value` might need to return a `Result`.